### PR TITLE
Add rpm spec

### PR DESCRIPTION
**Pull request by LecrisUT**
This will make it easier for maintainers to keep up to date with upstream package.

Note the version numbering is currently hard coded. It might be possible to retrieve it dynamically, but I'm not sure if it will work on copr. However this should not be much of an issue as long as we can get some notifications of the builds.

### CODE REVIEW

**Comment by LecrisUT** [comment]
In order to integrate with copr there are 2 ways, either we use my current copr repo setup, I give persmissions to the maintainers to manage it and we switch the hook to be called by this repo. Or one of the maintainers can create a copr repo (I can walk you through it if needed), and setup a webhook there. It is also possible to create groups, but this needs to be requested on the Fedora channels.

We can limit the webhook to tag creation, labels, etc. so it only triggers a build on new releases. In principle it should work, but I haven't tested this.

After that it is trivial for the maintainer to set a notification either on github or directly on copr. Need to check if they need builder permission to get notification on other's package for the latter.

**Comment by henryiii** [comment]
Is there an easy way I can try this locally (and therefore in Actions)?

FYI, rpmlint gives: `scikit-build-core.spec:6: W: mixed-use-of-spaces-and-tabs (spaces: line 6, tab: line 4)`

**Comment by LecrisUT** [comment]
The running on Actions is a bit tricky because in my experience with `mock` is that it gets a bit fussy when built locally with `ForgeMeta` (names of the tar ball created and expected name from the spec mismatch). I need to tinker a lot to find an actual workflow that works with this setup. Copr seems to use `rpkg` but I didn't look into how it manages to work.

Realistically though, even without local testing with actions, it would be a great help, especially since these builds should only be checked at release times.

**Comment by LecrisUT** [comment]
Ok, thanks to the folks from FedoraProject, I've figured `tito` is the best. Locally we can now build using:
```shell
$ tito build --srpm --test --output=srpm_dir
$ mock srpm_dir/name_of_srpm.srpm
```
Note that `tito` does not pick up uncommitted changes.

**Comment by LecrisUT** [comment]
We seem to need [this upstream PR](https://github.com/rpm-software-management/tito/pull/443) to be able to build using tito and vcs

**Comment by LecrisUT** [comment]
Ok, I got the spec to work, but it is not ideal. I think upstream `setuptools_scm` has a bug:
https://github.com/pypa/setuptools_scm/blob/f620057203f1446c1dea2ce57f73dccdd6f28ee9/src/setuptools_scm/git.py#L246-L250
That code errors on my side if it not the form `vXYZ-0-ABC`, so the `if len(split) < 3` is never triggered.

Could you confirm that with the `.git_archival.txt` from the tar ball the project is buildable? Anyway could you check the errors in the most [recent builds](https://copr.fedorainfracloud.org/coprs/lecris/scikit-build-core/package/scikit-build-core/)? [This one](https://download.copr.fedorainfracloud.org/results/lecris/scikit-build-core/fedora-38-x86_64/05537753-scikit-build-core/builder-live.log.gz) suggests that both `cmake` and `ninja-build` are required for running. The former is acceptable, but why can't we use a different build system for the latter?

**Comment by henryiii** [comment]
What form is it to cause this? The thing I see from looking at that is it's possibly missing the leading "v" stripping?

I can check the tarball.

We probably require cmake and ninja to run the tests. You don't need `ninja` to use scikit-build, `make` works fine too, it just the tests that likely require it. We probably can make the ninja-specific parts get skipped if ninja's not installed, but that probably hasn't been done, since GHA has ninja.

**Comment by LecrisUT** [comment]
So I have tested with both `v0.2.0` and `v0.2.0-1.git..`. It seems to require at least 2 dashes to be correctly picked up.

I have double-checked and redid the tests with `Requires: cmake`, but that didn't help. Seems to be a different issue. I misread the `no module cmake` are skipped so that was a red-hearing. Don't know what the real error is about though.

**Comment by henryiii** [comment]
```python
>>> "v0.2.0".rsplit("-", 2)
['v0.2.0']
```

Which has a length < 3, so that seems like it would pass. I'm guessing the leading v is stripped earlier.

**Comment by LecrisUT** [comment]
Oh, after a bit more debugging this was fixed in https://github.com/pypa/setuptools_scm/commit/4c2cf6e3a369afa05131d6fb3d790822b019abf0.

But that was introduced in `setuptools_scm v7.1.0`, and the Fedora 37 that I work on by default has `v7.0.5`. Sorry for the confusion on that.

Well the current version temporarily fixes it for F37, but it lacks other dependencies, so I think I'll drop it anyway.

There are still the build issues like
```
E               pytest_subprocess.exceptions.ProcessNotRegisteredError: The process '/usr/bin/cmake3 --version' was not registered.
```

Not sure how to debug these though, any ideas? `cmake3` works just fine in the mock environment:
```console
$ cmake3 --version
cmake version 3.26.0-rc4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

**Comment by LecrisUT** [comment]
Weird. The last error, when I manually run the tests, it doesn't fail, but for some reason it reliably fails when I run initially. Any insights around `test_pep517_sdist_hash`?

**Comment by henryiii** [comment]
Hmm, that's not a critical test, but it shouldn't be failing. That's checking that the SDist is binary reproducible. And you only have one of the tests failing. I'll see if I can come up with a way to detect what is different between the two files...

**Comment by henryiii** [comment]
What's the difference between "manually run the tests" and "run initially"?

**Comment by LecrisUT** [comment]
That's what puzzles me. Afaik there shouldn't be anything that affects it. If we can extract them, say maybe by binary dumping the files and reconstructing from that, maybe we can find out

**Comment by LecrisUT** [comment]
I have dumped the file as a base64 and tried to recreate it according to the latest temporary head commit.

[test.tar.gz](https://github.com/scikit-build/scikit-build-core/files/10862049/test.tar.gz)

**Comment by LecrisUT** [comment]
@henryiii so what do you think? Should this issue be fixed in the test, or should we disable this test and incorporate this as is?

**Comment by henryiii** [comment]
Dates should be locked when making SDists. This isn't a very important test, but it would be nice to solve.

**Comment by henryiii** [comment]
Is SOURCE_DATE_EPOCH set, perhaps? That would break this test, I think.

**Comment by henryiii** [comment]
Yeah, that's it, I see in the logs: `setting SOURCE_DATE_EPOCH=1677456000`. We can unset that for the tests that care about it.

File: tests/test_pyproject_pep517.py
Comment (henryiii) [inline tests/test_pyproject_pep517.py:74]:
```suggestion
    monkeypatch.delenv("SOURCE_DATE_EPOCH", raising=False)
```

**Comment by LecrisUT** [comment]
Indeed that was the case, thank you. I've cherrypicked the fix to a separate  PR, will rebase after.

File: tests/test_pyproject_pep517.py
Comment (LecrisUT) [reply tests/test_pyproject_pep517.py:74]: Ok I'll fix it in #221

**Comment by LecrisUT** [comment]
Let me know when `0.2.2` will be pushed, I'll start the process to submit it to FedoraProject after that.

**Comment by henryiii** [comment]
`0.2.2` is out.

**Comment by LecrisUT** [comment]
Thank you for the heads up. After a bit of battling with `copr` and `tito`, I managed to get it working.
https://copr.fedorainfracloud.org/coprs/lecris/fedora-scikit-build-core/builds/

**Comment by LecrisUT** [comment]
I was looking at the installed files. Can we remove `resources/find_python` since it is guaranteed to be present by the `cmake` dependency? Don't need to edit the installation or patch anything for that right?

**Comment by LecrisUT** [comment]
I was thinking about the discussion that was had on `scikit-build` about bundling files, and if that issue can arise again here. In principle I think it should be fine since we are not conflicting paths and searches, it is only used internally, but would be nice to keep track of such changes and clean up the repo every now and then. Still waiting on review comments so maybe it's a non-issue altogether.

File: scikit-build-core.spec
Comment (henryiii) [inline scikit-build-core.spec:16]: What version(s) of Python does this target? This requires `packaging` also, and possibly a a few other things on older Python versions? Or is there some reason that it doesn't need that?

**Comment by henryiii** [comment]
I'd still like some sort of smoke check, even if it's just a static analysis one that checks to see if the dependencies haven't changed and if the version is correct. I'm not going to remember to update the version - this is the only instance with a hardcoded version in the repo at the moment.

**Comment by LecrisUT** [comment]
Oh, it picked up some errors after rebasing onto main: https://download.copr.fedorainfracloud.org/results/lecris/scikit-build-core/fedora-38-aarch64/05686616-scikit-build-core/builder-live.log.gz

File: scikit-build-core.spec
Comment (LecrisUT) [reply scikit-build-core.spec:16]: So all of those are copied from `%pyproject_buildrequires`, except for the python version that is targeted. In that case it is purely which version is being released for that Fedora release. We just add/drop Fedora versions

**Comment by LecrisUT** [comment]
> I'd still like some sort of smoke check, even if it's just a static analysis one that checks to see if the dependencies haven't changed and if the version is correct. I'm not going to remember to update the version - this is the only instance with a hardcoded version in the repo at the moment.
> 
> But not a critical blocker if you want it in now.

I should be able to do that, but it would be a bit tricky to maintain because I haven't figured how to make dynamic version in the spec file. Currently we are running a workaround because of a missing feature in `tito`/`git archive`:
```
if grep -q "describe-name: \$Format" .git_archival.txt; then
   sed -i "s/describe-name:.*/describe-name: v%{version}/g" .git_archival.txt
fi
```

For upstream there is a different repo: https://src.fedoraproject.org/rpms/python-scikit-build-core/, because that one has to be built with fixed tar balls, while this one we want it to be built against latest version. This one should only be used for testing, i.e. a CI for fedora maintenance, so maybe having version difference will not be that critical. I'll still try to figure out how to make the version dynamic with fedora folks

**Comment by henryiii** [comment]
Are you using cmake 3.26.0? I think there's a bug if you have 3.26.0 exactly where it wasn't using the backported fix from the upcoming 3.26.1 - fixed in #238 (ahh, that's not even merged yet, that's probably why).

File: scikit-build-core.spec
Comment (henryiii) [reply scikit-build-core.spec:16]: Ahh, okay.

**Comment by LecrisUT** [comment]
Yep, that fixed it. [All green](https://copr.fedorainfracloud.org/coprs/lecris/scikit-build-core/build/5686765/)

**Review by henryiii — APPROVED** [review]

**Comment by henryiii** [comment]
(In theory, I think you can merge this)

**Comment by LecrisUT** [comment]
> (In theory, I think you can merge this)

Apparently needs more permission to write to protected branch

**Comment by henryiii** [comment]
Now?

**Comment by henryiii** [comment]
https://www.conventionalcommits.org/en/v1.0.0/ if possible.

**Pull request merged**
